File: src/main/java/main/BasePage.java
package main;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.ArrayList;
import java.util.Set;

public abstract class BasePage {
    protected WebDriver driver;
    protected WebDriverWait webDriverWait;
    private String title;

    public BasePage(WebDriver driver){
        this.driver = driver;
        webDriverWait = new WebDriverWait(driver, 3);
        this.title = title;
    }

    public void switchToNewWindow(String mainWindowHandle) {
        Set<String> allWindowHandles = driver.getWindowHandles();
        for (String windowHandle : allWindowHandles) {
            if (!windowHandle.equals(mainWindowHandle)) {
                driver.switchTo().window(windowHandle);
                break;
            }
        }
    }
    public void switchToLastTab() {
        ArrayList<String> tabs = new ArrayList<>(driver.getWindowHandles());
        try {
            driver.switchTo().window(tabs.get(tabs.size() - 1));
        } catch (IndexOutOfBoundsException ex) {
            throw new RuntimeException("Switching to tab is impossible because it doesn't exist.");
        }
    }

    public void switchToFrame(WebElement frameElement) {
        driver.switchTo().frame(frameElement);
    }

    public void switchToParentFrame() {
        driver.switchTo().parentFrame();
    }

    public void switchToNewWindowTest() {
        for (String winHandle : driver.getWindowHandles()) {
            driver.switchTo().window(winHandle);
        }
    }
}

File: src/main/java/main/BasePage.kt
import org.openqa.selenium.WebDriver
import org.openqa.selenium.WebElement
import org.openqa.selenium.chrome.ChromeDriver
import org.openqa.selenium.support.ui.WebDriverWait

abstract class BasePage(protected var driver: WebDriver?) {
    protected var webDriverWait: WebDriverWait = WebDriverWait(driver ?: throw IllegalArgumentException("WebDriver must not be null"), 3)

    open fun switchToNewWindow(mainWindowHandle: String) {
        driver?.let {
            val allWindowHandles = it.windowHandles
            for (windowHandle in allWindowHandles) {
                if (windowHandle != mainWindowHandle) {
                    it.switchTo().window(windowHandle)
                    break
                }
            }
        }
    }

    fun switchToLastTab() {
        driver?.let {
            val tabs = ArrayList(it.windowHandles)
            try {
                it.switchTo().window(tabs[tabs.size - 1])
            } catch (ex: IndexOutOfBoundsException) {
                throw RuntimeException("Switching to tab is impossible because it doesn't exist.")
            }
        }
    }

    fun switchToFrame(frameElement: WebElement?) {
        driver?.switchTo()?.frame(frameElement)
    }

    fun switchToParentFrame() {
        driver?.switchTo()?.parentFrame()
    }

    fun switchToNewWindowTest() {
        driver?.let {
            for (winHandle in it.windowHandles) {
                it.switchTo().window(winHandle)
            }
        }
    }
}


